Ensure `lld` is supported with `download-ci-llvm`

lqd opened the pull request:
This PR:
- ensures LLD's step in bootstrap's dist, but it's not strictly necessary since dist will already package it when it's present.
- makes bootstrap's `native::LLD` step support using the packaged `ci-llvm/bin/lld`, instead of building it from source (which would most likely not be available today, nor in the future where `download-ci-llvm = if-available` is the default).

If I understand correctly, `--enable-full-tools` will also enable `rust.lld`, and this is why LLD is already packaged today in the `rust-dev` component on the main targets (and why `-Zgcc-ld=lld` does work there).

That means it's likely that this PR will not be able to land before I've reworked and landed #101792: if LLD is available in `download-ci-llvm`, the `needs-rust-lld` tests should start being executed on the x64 macOS test builders, and CI would fail today.

I've tested locally that building with `download-ci-llvm = true` and `lld = true` with the LLVM submodule unregistered was successful, and that `rust-lld` and the various `lld-wrapper`s are present and `-Zgcc-ld=lld` works as well, on a few different platforms:
- `x86_64-unknown-linux-gnu`
- `aarch64-apple-darwin`
- `x86_64-pc-windows-msvc` (with `-Clinker=rust-lld` rather than `-Zgcc-ld=lld`)
- `x86_64-apple-darwin`, with the `MACOSX_DEPLOYMENT_TARGET` workaround for #101653

I don't think we really need to bump the `download-ci-llvm-stamp` in this case, since `./build/$triple/ci-llvm/bin/lld` is present on all the above targets already, but have added it mechanically, and it should probably be removed to avoid unnecessary downloads/churn.

Fixes #98340
Supersedes #100010

lqd committed:
bump download-ci-llvm-stamp

lqd commented:
@bors try

bors commented:
:hourglass: Trying commit fdf2238424e42ef67cd1826723a792c86735ded7 with merge 2b02e73d38e5508c9eeda99c5c321b11cd0667f6...
<!-- homu: {"type":"TryBuildStarted","head_sha":"fdf2238424e42ef67cd1826723a792c86735ded7","merge_sha":"2b02e73d38e5508c9eeda99c5c321b11cd0667f6"} -->

lqd committed:
make the `native::LLD` step able to use an already built lld

Makes the lld step avoid building it from source when possible: when
dist has packaged it along the other LLVM binaries for the rust-dev
component.

lqd committed:
slight cleanup

lqd committed:
turn a comment into an actual doc comment

lqd committed:
bootstrap dist: ensure LLD's step

jyn514 commented on src/bootstrap/dist.rs:
```diff
@@ -2058,6 +2058,12 @@ impl Step for RustDev {
 
         builder.ensure(crate::native::Llvm { target });
 
+        // If the config has LLD enabled, ensure its step. We'll we want to package it, and use it
+        // in download-ci-llvm.
+        if builder.config.lld_enabled {
+            builder.ensure(crate::native::Lld { target });
```
Shouldn't we do this regardless of whether lld is enabled? We want `RustDev` to be the same across all platforms.

lqd replied:
Do you mean the contents of the archive (as this is already not the same across all platforms IIUC) ?

jyn514 replied:
Yes, I mean the contents.

>  But as mentioned in the OP, this ensure is not strictly necessary, packaging lld is already done if it's found on disk.

Right, and I think that's very non-intuitive and we should change it. `lld` should be packaged unconditionally, not based on `lld_enabled` - that config says
> Indicates whether LLD will be compiled and made available in the sysroot for rustc to execute.

but IMO that should only affect whether LLD is *used* when building rustc, not whether we package it or not.

What would be the purpose of making this configurable? We'll want to make sure it's enabled in all dist builders anyway, and at that point it just becomes one more thing we need to test.

lqd replied:
> but IMO that should only affect whether LLD is _used_ when building rustc

There's another different setting for that, `rust.use-lld`. I assumed the purpose of `rust.lld` is to be able to choose the subset of platforms for which we distribute `rust-lld`, e.g. maybe it itself doesn't support some platforms we need, but I don't know either. 

If the comment `// We don't build LLD on some platforms, so only add it if it exists` is still true, it may not yet be possible to package `lld` unconditionally. We'll probably have to see what CI has to say about that :)

jyn514 replied:
god, what a mess :(

ok, we don't need to dig into the difference here. But I do still think `RustDev` should always have lld in it.

jyn514 commented on src/bootstrap/native.rs:
```diff
@@ -803,7 +803,21 @@ impl Step for Lld {
-        let llvm_config = builder.ensure(Llvm { target: self.target });
+        let llvm_config = builder.ensure(Llvm { target });
+
+        // The `dist` step packages LLD next to LLVM's binaries for download-ci-llvm. The root path
+        // we usually expect here is `./build/$triple/ci-llvm/`, with the binaries in its `bin`
+        // subfolder. We check if that's the case, and if LLD's binary already exists there next to
+        // `llvm-config`: if so, we can use it instead of building LLVM/LLD from source.
+        let ci_llvm_bin = llvm_config.parent().unwrap();
+        if ci_llvm_bin.is_dir() && ci_llvm_bin.file_name().unwrap() == "bin" {
+            let lld_path = ci_llvm_bin.join(exe("lld", target));
+            if lld_path.exists() {
+                // The following steps copying `lld` as `rust-lld` to the sysroot, expect it in the
+                // `bin` subfolder of this step's out dir.
+                return ci_llvm_bin.parent().unwrap().to_path_buf();
```
Why does LLD need to be special cased this way? It looks like the `Llvm` step just builds unconditionally. Should we put this logic in the `dist` step instead of here?

lqd replied:
This logic already is in the `dist` step for the [`RustDev` component](https://github.com/rust-lang/rust/blob/8de4b138455add55bde6de5553a933a2ab79b71f/src/bootstrap/dist.rs#L2081-L2085).  `dist` will optionally package `lld` in the `RustDev` component if it can find it on disk, but the `native::Lld` step is always expecting to build it from source. 

However, the `native::Llvm` step doesn't [just build unconditionally](https://github.com/rust-lang/rust/blob/8de4b138455add55bde6de5553a933a2ab79b71f/src/bootstrap/native.rs#L255-L259), it [can work with `download-ci-llvm`](https://github.com/rust-lang/rust/blob/8de4b138455add55bde6de5553a933a2ab79b71f/src/bootstrap/native.rs#L68-L77) and this change is the dual to those, making `native::Lld` able to read what `dist` already writes to the component.

jyn514 replied:
> However, the native::Llvm step doesn't [just build unconditionally](https://github.com/rust-lang/rust/blob/8de4b138455add55bde6de5553a933a2ab79b71f/src/bootstrap/native.rs#L255-L259), it [can work with download-ci-llvm](https://github.com/rust-lang/rust/blob/8de4b138455add55bde6de5553a933a2ab79b71f/src/bootstrap/native.rs#L68-L77) and this change is the dual to those, making native::Lld able to read what dist already writes to the component.

:+1: makes sense, thanks for explaining :)

jyn514 reviewed (commented):
> If I understand correctly, --enable-full-tools will also enable rust.lld, and this is why LLD is already packaged today in the rust-dev component on the main targets (and why -Zgcc-ld=lld does work there).

Can you expand on "main targets"? Is it available on all tier 1 targets, or only targets where we build LLVM from source? and if it's on all tier 1 targets, what is this change doing?

jyn514 reviewed (approved):
Thanks for explaining. This looks good to me if you change LLD to be included in `RustDev` unconditionally.

jyn514 commented:
> That means it's likely that this PR will not be able to land before I've reworked and landed https://github.com/rust-lang/rust/pull/101792: if LLD is available in download-ci-llvm, the needs-rust-lld tests should start being executed on the x64 macOS test builders, and CI would fail today.

I am not quite sure what the relation is here, but FWIW I don't have any problem with speculatively approving this and seeing if CI fails or not.

lqd commented:
To summarize the intent I described on zulip: our `rust-lld` support (e.g. `-Zgcc-ld=lld`) has been broken during refactoring a couple times and CI didn't catch it: my understanding is that no test builders are running the existing `needs-rust-lld` tests (they don't build llvm/lld from source), and lld is only enabled on the dist builders. 

This PR is ultimately to fix CI so that this doesn't regress anymore: the test builders will have access to lld with `download-ci-llvm`. But since things are still currently broken on x64 apple, I'll have to fix that before this PR lands fixing CI: the `needs-rust-lld` tests would automatically be enabled, and fail because of the breakage.

bors commented:
:umbrella: The latest upstream changes (presumably #106000) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lqd committed:
ensure lld's step unconditionally for RustDev component

jyn514 commented:
Thanks, this looks great!

> I'm not sure we really need to bump the `download-ci-llvm-stamp` here (since the lld binary should already be in `RustDev` if lld is enabled on the builder), let me know if you want me to remove that ?

I think it's not strictly necessary, but there's no harm bumping it, the download isn't too big.

@bors r+

bors commented:
:pushpin: Commit e0f5c6da1df52a64b80753bf33bae3176729ae52 has been approved by `jyn514`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jyn514 e0f5c6da1df52a64b80753bf33bae3176729ae52 -->
<!-- homu: {"type":"Approved","sha":"e0f5c6da1df52a64b80753bf33bae3176729ae52","approver":"jyn514","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
